File: VB/Reporting_how-to-override-the-existing-behavior-when-dropping-an-item-from-the-field-list-in-e478/Form1.vb
Imports System
Imports System.Windows.Forms
Imports DevExpress.XtraReports.Design
Imports DevExpress.XtraReports.UserDesigner

Namespace FieldListDragDrop
	Partial Public Class Form1
		Inherits Form

		Public Sub New()
			InitializeComponent()
		End Sub

        Private Sub ShowReportDesigner()
            Dim designForm As New XRDesignFormEx()

            AddHandler designForm.DesignPanel.DesignerHostLoaded, AddressOf DesignPanel_DesignerHostLoaded

            designForm.OpenReport(New XtraReport1())

            designForm.ShowDialog()
        End Sub

        Private Sub DesignPanel_DesignerHostLoaded(ByVal sender As Object, ByVal e As DesignerLoadedEventArgs)
			e.DesignerHost.RemoveService(GetType(IFieldListDragDropService))
			e.DesignerHost.AddService(GetType(IFieldListDragDropService), New MyFieldListDragDropService(e.DesignerHost))
		End Sub

        Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load
            ShowReportDesigner()
        End Sub
    End Class

End Namespace

File: CS/Reporting_how-to-override-the-existing-behavior-when-dropping-an-item-from-the-field-list-in-e478/Form1.cs
using System;
using System.Windows.Forms;
using DevExpress.XtraReports.Design;
using DevExpress.XtraReports.UserDesigner;

namespace FieldListDragDrop {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }

        private void ShowReportDesigner() { 
            XRDesignFormEx designForm = new XRDesignFormEx();

            designForm.DesignPanel.DesignerHostLoaded +=
                new DesignerLoadedEventHandler(DesignPanel_DesignerHostLoaded);

            designForm.OpenReport(new XtraReport1());

            designForm.ShowDialog();
        }

        void DesignPanel_DesignerHostLoaded(object sender, DesignerLoadedEventArgs e) {
            e.DesignerHost.RemoveService(typeof(IFieldListDragDropService));
            e.DesignerHost.AddService(typeof(IFieldListDragDropService), 
                new MyFieldListDragDropService(e.DesignerHost));
        }

        private void Form1_Load(object sender, EventArgs e) {
            ShowReportDesigner();
        }
    }

}
